File: pe/visualstudio/PE.FSharp.Tests/Problem0011.Tests.fs
﻿namespace PE.Problem0011.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open PE.FSharp.Problem11.Problem0011

module Problem11Tests = 
  
    [<TestClass>]
    type Problem0011() = 

        let testSpace1Path = "..\..\..\..\..\..\practice\pe\problemsets\problem0011test1.txt"
        let testSpace2Path = "..\..\..\..\..\..\practice\pe\problemsets\problem0011test2.txt"
        let questionSpacePath = "..\..\..\..\..\..\practice\pe\problemsets\problem0011.txt"

        [<TestMethod>]
        member x.``Problem0011 importProblemSpace testSpace1Path should return testSpace1.``() = 
            Assert.AreEqual([[0;0;0;0;1];[1;1;2;2;3];[1;1;4;4;5];[2;5;6;7;9];[3;9;9;9;8]], importProblemSpace testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 importProblemSpace testSpace2Path should return testSpace2.``() = 
            let result = importProblemSpace testSpace2Path
            Assert.AreEqual([[8;9;9;9;9];[1;1;2;2;9];[1;1;4;4;9];[2;5;6;7;9];[3;9;9;9;8]], result)

        [<TestMethod>]
        member x.``Problem0011 maxOfHorizontalMaxProducts should find the largest product of the test space 1 on the horizontal lines. In this case: 5832.``() =
            Assert.AreEqual(5832, maxOfHorizontalMaxProducts testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 maxOfHorizontalMaxProducts should find the largest product of the test space 2 on the horizontal lines. In this case: 6561.``() =
            Assert.AreEqual(6561, maxOfHorizontalMaxProducts testSpace2Path)

        [<TestMethod>]
        member x.``Problem0011 constructVerticalTuples should return correct list of tuples from testSpace1.``() =
            Assert.AreEqual(
                [
                    (0,1,1,2);
                    (0,1,1,5);
                    (0,2,4,6);
                    (0,2,4,7);
                    (1,3,5,9);
                    (1,1,2,3);
                    (1,1,5,9);
                    (2,4,6,9);
                    (2,4,7,9);
                    (3,5,9,8)
                ], constructVerticalTuples testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 constructDiagonalTuples should return correct list of tuples from testSpace1.``() =
            Assert.AreEqual(
                [
                    (0,1,4,7);
                    (0,2,4,9);
                    (1,1,6,9);
                    (1,4,7,8)
                ], constructDiagonalTuples testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 constructDiagonalTuplesLToR should return correct list of tuples from testSpace1.``() =
            Assert.AreEqual(
                maxOfListOf4WideTuples
                    [
                        (1,2,4,5);
                        (0,2,1,2);
                        (3,4,6,9);
                        (2,4,5,3)
                    ],
                maxOfListOf4WideTuples (constructDiagonalTuplesLToR testSpace1Path))

        [<TestMethod>]
        member x.``Problem0011 maxOfAllVerticalAndDiagonalTuples on testSpace1 should return 1080.``() =
            Assert.AreEqual(1080, maxOfAllVerticalAndDiagonalTuples testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 maxOfAllVerticalAndDiagonalTuples on testSpace2 should return 6561.``() =
            Assert.AreEqual(6561, maxOfAllVerticalAndDiagonalTuples testSpace2Path)

        [<TestMethod>]
        member x.``Problem0011 maxOfListOf4WideTuples on testSpace1 should return 240.``() =
            Assert.AreEqual(240, maxOfListOf4WideTuples [(2,2,2,30);(1,2,3,4);(1,2,3,4)])

        [<TestMethod>]
        member x.``Problem0011 maxOfListOf4WideTuples on testSpace2 should return 10000.``() =
            Assert.AreEqual(10000, maxOfListOf4WideTuples [(2,2,2,30);(1,2,3,4);(1,2,3,4);(10,10,10,10)])

        [<TestMethod>]
        member x.``Problem0011 maxOfAllSetsOfFour on testSpace1 should return 5832.``() =
            Assert.AreEqual(5832, maxOfAllSetsOfFour testSpace1Path)

        [<TestMethod>]
        member x.``Problem0011 maxOfAllSetsOfFour on testSpace2 should return 6561.``() =
            Assert.AreEqual(6561, maxOfAllSetsOfFour testSpace2Path)

        [<TestMethod>]
        member x.``Problem0011 answer``() =
            Assert.AreEqual(70600674, maxOfAllSetsOfFour questionSpacePath)

File: pe/visualstudio/PE.CSharp.Tests/Problem0011.Tests.cs
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace PE.CSharp.Tests
{
    [TestClass]
    public class Problem0011Tests
    {
        private static Problem0011 _problem0011;
        private const string TestSpace1Path = @"c:\Users\Thomas\Documents\GitHub\practice\pe\problemsets\problem0011test1.txt";
        private const string TestSpace2Path = @"c:\Users\Thomas\Documents\GitHub\practice\pe\problemsets\problem0011test2.txt";
        private const string QuestionSpacePath = @"c:\Users\Thomas\Documents\GitHub\practice\pe\problemsets\problem0011.txt";
        private static List<List<int>> _test1;
        private static List<List<int>> _test2;

        [ClassInitialize]
        public static void InitializeClass(TestContext ctx)
        {
            _problem0011 = new Problem0011();
            _test1 = new List<List<int>>
            {
                new List<int> {0, 0, 0, 0, 1},
                new List<int> {1, 1, 2, 2, 3},
                new List<int> {1, 1, 4, 4, 5},
                new List<int> {2, 5, 6, 7, 9},
                new List<int> {3, 9, 9, 9, 8}
            };

            _test2 = new List<List<int>>
            {
                new List<int> {8, 9, 9, 9, 9},
                new List<int> {1, 1, 2, 2, 9},
                new List<int> {1, 1, 4, 4, 9},
                new List<int> {2, 5, 6, 7, 9},
                new List<int> {3, 9, 9, 9, 8}
            };

        }

        [TestMethod]
        public void Problem0011_ImportProblemSpace_Test1_Should_Return_Correct_ProblemSpace()
        {
            ImportProblemSpaceAssert(_test1, TestSpace1Path);
        }

        [TestMethod]
        public void Problem0011_ImportProblemSpace_Test2_Should_Return_Correct_ProblemSpace()
        {
            ImportProblemSpaceAssert(_test2, TestSpace2Path);
        }

        private void ImportProblemSpaceAssert(IReadOnlyList<List<int>> expected, string path)
        {
            var result = _problem0011.ImportProblemSpace(path);

            for (var i = 0; i < expected.Count; i++)
            {
                Assert.IsTrue(expected[i].SequenceEqual(result[i]));
            }
        }

        [TestMethod]
        public void Problem0011_MaxProductOfHorizontals_Test1_Should_Return_5832()
        {
            MaxProductOfHorizontalsAssert(5832, _test1);
        }

        [TestMethod]
        public void Problem0011_MaxProductOfHorizontals_Test2_Should_Return_6561()
        {
            MaxProductOfHorizontalsAssert(6561, _test2);
        }

        private void MaxProductOfHorizontalsAssert(int expected ,List<List<int>> testData)
        {
            Assert.AreEqual(expected, _problem0011.MaxProductOfHorizontals(testData));
        }

        [TestMethod]
        public void Problem0011_MaxProductOfVerticles_Test1_Should_Return_1080()
        {
            MaxProductOfVerticlesAssert(1080, _test1);
        }

        [TestMethod]
        public void Problem0011_MaxProductOfVerticles_Test2_Should_Return_6561()
        {
            MaxProductOfVerticlesAssert(6561, _test2);
        }

        private void MaxProductOfVerticlesAssert(int expected, List<List<int>> testData)
        {
            Assert.AreEqual(expected, _problem0011.MaxProductOfVerticles(testData));
        }

        [TestMethod]
        public void Problem0011_MaxProductOfForwardDiagonals_Test1_Should_Return_648()
        {
            MaxProductOfForwardDiagonalsAssert(648, _test1);
        }

        [TestMethod]
        public void Problem0011_MaxProductOfForwardDiagonals_Test2_Should_Return_1944()
        {
            MaxProductOfForwardDiagonalsAssert(1944, _test2);
        }

        private void MaxProductOfForwardDiagonalsAssert(int expected, List<List<int>> testData)
        {
            Assert.AreEqual(expected, _problem0011.MaxProductOfForwardDiagonals(testData));
        }

        [TestMethod]
        public void Problem0011_MaxProductOfBackwardDiagonals_Test1_Should_Return_224()
        {
            MaxProductOfBackwardDiagonalsAssert(224, _test1);
        }

        [TestMethod]
        public void Problem0011_MaxProductOfBackwardDiagonals_Test2_Should_Return_648()
        {
            MaxProductOfBackwardDiagonalsAssert(648, _test2);
        }

        private void MaxProductOfBackwardDiagonalsAssert(int expected, List<List<int>> testData)
        {
            Assert.AreEqual(expected, _problem0011.MaxProductOfBackwardDiagonals(testData));
        }

        [TestMethod]
        public void Problem0011_MaxProductOf4ConnectedNumbers_Test1Path_Should_Return_5832()
        {
            MaxProductOf4ConnectedNumbersAssert(5832, TestSpace1Path);
        }

        [TestMethod]
        public void Problem0011_MaxProductOf4ConnectedNumbers_Test2Path_Should_Return_6561()
        {
            MaxProductOf4ConnectedNumbersAssert(6561, TestSpace2Path);
        }

        [TestMethod]
        public void Problem0011_MaxProductOf4ConnectedNumbers_ProblemSpace_Should_Return_70600674()
        {
            MaxProductOf4ConnectedNumbersAssert(70600674, QuestionSpacePath);
        }

        private void MaxProductOf4ConnectedNumbersAssert(int expected, string inputPath)
        {
            Assert.AreEqual(expected, _problem0011.MaxProductOf4ConnectedNumbers(inputPath));
        }
    }
}
